src/coreclr/jit/morphblock.cpp

```diff
@@ -0,0 +1,1551 @@
+    DISPTREE(tree);
+    return tree;
+}
+
+//------------------------------------------------------------------------
+// MorphCommaBlock: transform COMMA<struct>(X) as OBJ<STRUCT>(COMMA byref(ADDR(X)).
+//
+// Notes:
+//    In order to CSE and value number array index expressions and bounds checks,
+//    the commas in which they are contained need to match.
+//    The pattern is that the COMMA should be the address expression.
+//    Therefore, we insert a GT_ADDR just above the node, and wrap it in an obj or ind.
+//    TODO-1stClassStructs: Consider whether this can be improved.
+//    Example:
+//      before: [3] comma struct <- [2] comma struct <- [1] LCL_VAR struct
+//      after: [5] obj <- [3] comma byref <- [2] comma byref <- [4] addr byref <- [1] LCL_VAR struct
+//
+// static
+GenTree* MorphInitBlockHelper::MorphCommaBlock(Compiler* comp, GenTreeOp* firstComma)
+{
+    assert(firstComma->OperIs(GT_COMMA));
+
+    Compiler::GenTreePtrStack commas(comp->getAllocator(CMK_ArrayStack));
+    for (GenTree* currComma = firstComma; currComma != nullptr && currComma->OperIs(GT_COMMA);
+         currComma          = currComma->gtGetOp2())
+    {
+        commas.Push(currComma);
+    }
+
+    GenTree* lastComma = commas.Top();
+
+    GenTree* effectiveVal = lastComma->gtGetOp2();
+
+    assert(effectiveVal == firstComma->gtEffectiveVal());
+
+    GenTree* effectiveValAddr = comp->gtNewOperNode(GT_ADDR, TYP_BYREF, effectiveVal);
+
+    INDEBUG(effectiveValAddr->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);
+
+    lastComma->AsOp()->gtOp2 = effectiveValAddr;
+
+    while (!commas.Empty())
+    {
+        GenTree* comma = commas.Pop();
+        comma->gtType  = TYP_BYREF;
+        comp->gtUpdateNodeSideEffects(comma);
+    }
+
+    const var_types blockType = effectiveVal->TypeGet();
+    GenTree*        addr      = firstComma;
+
+    GenTree* res;
+
+    if (blockType == TYP_STRUCT)
+    {
+        CORINFO_CLASS_HANDLE structHnd = comp->gtGetStructHandleIfPresent(effectiveVal);
+        if (structHnd == NO_CLASS_HANDLE)
+        {
+            // TODO-1stClassStructs: get rid of all such cases.
+            res = comp->gtNewIndir(blockType, addr);
+        }
+        else
+        {
+            res = comp->gtNewObjNode(structHnd, addr);
+            comp->gtSetObjGcInfo(res->AsObj());
+        }
+    }
+    else
+    {
+        res = comp->gtNewIndir(blockType, addr);
+    }
+
+    comp->gtUpdateNodeSideEffects(res);
+    INDEBUG(res->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);
+    return res;
+}
+
+//------------------------------------------------------------------------
+// MorphDynamicBlock: tries to transform a dynamic block as a const block.
+//
+// static
+GenTreeBlk* MorphInitBlockHelper::MorphDynamicBlock(Compiler* comp, GenTreeDynBlk* dynBlock)
+{
+    if (dynBlock->gtDynamicSize->IsCnsIntOrI())
+    {
+        GenTreeIntCon* dynSize = dynBlock->gtDynamicSize->AsIntCon();
+        assert(dynSize->FitsInI32());
```

Comment:
Noting that this is an interesting assert in that it is actually testing that the constant node itself is well-formed. The implicit contract of `CNS_INT` of `TYP_INT` on 64 bit hosts is that the `int32_t` must be sign-extended to `int64_t`, which means that this assert will fail if something (say buggy `GT_CAST` folding...) creates `dynSize` with `gtIconValue` that is actually equal to `UINT32_MAX` (i. e. has upper bits set to zero), even as `MorphDynamicBlock` itself can work with such a constant (it won't be actually truncating anything).